Fix statistic counters for parent memory contexts when some child one attaches/detaches

red1452 opened the pull request:
Counters of memory contexts became incorrect when we deleted child memory context from it. At first pointer to parent memory context was reset, at second all allocated memory was freed and counters were updated. Because of reseted pointer to parent memory context, counters of parent memory context and their parents are not updated.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

RekGRpth commented:
may be better is sync code to upstream master (gpdb 7) ?

RekGRpth commented on src/backend/utils/mmgr/aset.c:
```diff
@@ -262,7 +262,7 @@ static void AllocSetFreeHeader(MemoryContext context, void *pointer);
 static void AllocSetInit(MemoryContext context);
 static void AllocSetReset(MemoryContext context);
-static void AllocSetDelete(MemoryContext context);
+static Size AllocSetDelete(MemoryContext context);
```
I think it is not good idea to change postgres api

red1452 replied:
I do it like it is at greenplum master.

RekGRpth commented:
You should add some test, I think.

red1452 commented:
> You should add some test, I think.

I think about it, but I do not know how make it.

RekGRpth commented:
> > You should add some test, I think.
> 
> I think about it, but I do not know how make it.

like in reproduce

red1452 commented:
> > > You should add some test, I think.
> > 
> > 
> > I think about it, but I do not know how make it.
> 
> like in reproduce

Look carefully, to see the problem, you need to connect using gdb and need to execute `p MemoryContextStats(TopMemoryContext)` at gdb.

RekGRpth commented on src/backend/utils/mmgr/aset.c:
```diff
@@ -1109,6 +1109,7 @@ AllocSetDelete(MemoryContext context)
 		AllocBlock	next = block->next;
 		size_t freesz = UserPtr_GetUserPtrSize(block);
         MemoryContextNoteFree(&set->header, freesz);
+		MemoryContextNoteFree(parent, freesz);
```
why not
```c
AllocSet pset = (AllocSet) parent;
AssertArg(AllocSetIsValid(pset));
MemoryContextNoteFree(&pset->header, freesz);
```

maksm90 replied:
May be is it worth to inject `MemoryContextNoteFree()` call for parent context into `MemoryContextDeleteImpl()` before recursive `MemoryContextDeleteChildren()` call? And call for current context (within `AllocSetDelete()`) after its delinking completely remove as this operation is meaningfulless on current stage?

maksm90 replied:
smth like this:
```C
diff --git a/src/backend/utils/mmgr/aset.c b/src/backend/utils/mmgr/aset.c
index 4cde8a874c..26739403af 100644
--- a/src/backend/utils/mmgr/aset.c
+++ b/src/backend/utils/mmgr/aset.c
@@ -1107,8 +1107,6 @@ AllocSetDelete(MemoryContext context)
     while (block != NULL)
     {
         AllocBlock  next = block->next;
-        size_t freesz = UserPtr_GetUserPtrSize(block);
-        MemoryContextNoteFree(&set->header, freesz);
 
 #ifdef CLOBBER_FREED_MEMORY
         wipe_mem(block, block->freeptr - ((char *) block));
diff --git a/src/backend/utils/mmgr/mcxt.c b/src/backend/utils/mmgr/mcxt.c
index de22f54005..a614c77ed5 100644
--- a/src/backend/utils/mmgr/mcxt.c
+++ b/src/backend/utils/mmgr/mcxt.c
@@ -228,6 +228,8 @@ MemoryContextResetChildren(MemoryContext context)
 void
 MemoryContextDeleteImpl(MemoryContext context, const char* sfile, const char *func, int sline)
 {
+    size_t freesz = context->allBytesAlloc - context->allBytesFreed;
+
     AssertArg(MemoryContextIsValid(context));
     /* We had better not be deleting TopMemoryContext ... */
     Assert(context != TopMemoryContext);
@@ -239,6 +241,7 @@ MemoryContextDeleteImpl(MemoryContext context, const char* sfile, const char *fu
     context->callerLine = sline;
 #endif
 
+    MemoryContextNoteFree(context->parent, freesz);
     MemoryContextDeleteChildren(context);
 
     /*
```

red1452 committed:
Fix statistic counters for parent memory contexts when some child one attaches/detaches (#404)

Counters of memory contexts became incorrect when we deleted child memory context from it.
At first pointer to parent memory context was reset, at second all allocated memory was freed
and counters were updated. Because of reseted pointer to parent memory context, counters of
parent memory context and their parents are not updated.

maksm90 commented on src/backend/utils/mmgr/aset.c:
```diff
@@ -1117,6 +1118,8 @@ AllocSetDelete(MemoryContext context)
 		block = next;
 	}
 
+	MemoryContextNoteFree(&set->header, allFreed);
```
I think if we believe that `context->allBytesAlloc - context->allBytesFreed` is equaled to the sum of sizes of allocated blocks we have to use this formula to exclude redundant computation in the body of loop.

maksm90 reviewed (dismissed):
Good catch!
Pls revise the description of problem, smth like "Fix statistic counters for parent memory contexts when some child one attaches/detaches"
